include/picongpu/particles/collision/binary/DefaultAlg.hpp

```diff
@@ -0,0 +1,652 @@
+                            float3_COLL const labMomentum1
+                                = precisionCast<float_COLL>(par1[momentum_]) / normalizedWeight1;
+                            float_COLL const mass0 = precisionCast<float_COLL>(picongpu::traits::attribute::getMass(
+                                particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE,
+                                par0));
+                            float_COLL const mass1 = precisionCast<float_COLL>(picongpu::traits::attribute::getMass(
+                                particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE,
+                                par1));
+                            float_COLL const charge0
+                                = precisionCast<float_COLL>(picongpu::traits::attribute::getCharge(
+                                    particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE,
+                                    par0));
+                            float_COLL const charge1
+                                = precisionCast<float_COLL>(picongpu::traits::attribute::getCharge(
+                                    particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE,
+                                    par1));
+                            float_COLL const gamma0 = picongpu::gamma<float_COLL>(labMomentum0, mass0);
+                            float_COLL const gamma1 = picongpu::gamma<float_COLL>(labMomentum1, mass1);
+
+                            //                            PMACC_DEVICE_ASSERT_MSG(
+                            //                                math::sqrt(pmacc::math::abs2(labMomentum0)) / gamma0 /
+                            //                                mass0 < c, "labVelocity0: %e",
+                            //                                sqrt(pmacc::math::abs2(labMomentum0)) / gamma0 / mass0);
+                            //                            PMACC_DEVICE_ASSERT_MSG(
+                            //                                math::sqrt(pmacc::math::abs2(labMomentum1)) / gamma1 /
+                            //                                mass1 < c, "labVelocity1: %e",
+                            //                                sqrt(pmacc::math::abs2(labMomentum1)) / gamma1 / mass1);
+                            // [Perez 2012] (1)
+                            // todo:: maybe this also with doubles?
+                            float3_COLL const comsVelocity
+                                = (labMomentum0 + labMomentum1) / (mass0 * gamma0 + mass1 * gamma1);
+                            float_COLL const comsVelocityAbs2 = pmacc::math::abs2(comsVelocity);
+                            float3_COLL comsMomentum0;
+                            float_COLL gammaComs, factorA, coeff0, coeff1;
+
+                            if(comsVelocityAbs2 != 0.0_COLL)
+                            {
+                                gammaComs = 1.0_COLL / math::sqrt(1.0_COLL - comsVelocityAbs2 / c / c);
```

Comment:
Discussed in VC regarding potential floating-point issues in this part of the code.

- Ideally, if possible it would be better to rewrite these equations to use momentums, not velocities, and use standard `picongpu::gamma()` to compute gamma from momentum as it should have no numerical issues
- As a quick potential fix to this expression (and similar one elsewhere if you have them) is to replace `1.0_COLL - comsVelocityAbs2 / c / c` with `(1.0_COLL - comsVelocityAbs2 / c) * (1.0_COLL + comsVelocityAbs2 / c)`. It should introduce no additional issues except a couple more operations, and be much better in case v is close to c. So that's a more safe way to write it essentially for free while you are debugging

Reply:
@sbastrakov  I don't really understand why it's better when I have to calculate `comsVelocityAbs` with `sqrt(comsVelocityAbs2)` anyway but I trust you with this.